### PR TITLE
change backend to deployment for serve

### PR DESCRIPTION
**Pull request by jiaodong**
This makes serve terms clear without "backend" variables hanging around.

A few things to pay attention to:
 - Java files are also changed, but corresponding protobuf def are modified too
 - Not sure what this means to product side yet, need to wait for tests to finish

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by jiaodong** [comment]
@simon-mo can you help to verify Ant / Java related changes are safe ? @architkulkarni  for snapshots, runtime env or product integration related pieces. Thanks !

File: python/ray/serve/router.py [hunk 1]
```diff
@@ -200,7 +201,7 @@ def __init__(
         self.long_poll_client = LongPollClient(
             controller_handle,
             {
-                (LongPollNamespace.BACKEND_CONFIGS, backend_tag): self.
+                (LongPollNamespace.deployment_configS, backend_tag): self.
```
Comment (architkulkarni) [inline python/ray/serve/router.py:204]:
```suggestion
                (LongPollNamespace.DEPLOYMENT_CONFIGS, backend_tag): self.
```

**Review by architkulkarni — APPROVED** [review]
snapshot changes look fine, since the content of the snapshot doesn't appear to be changed by the PR (and this is unit-tested too.). Also I don't see anything touching runtime envs here, so overall looks good to me!

File: java/serve/src/main/java/io/ray/serve/RayServeReplica.java [hunk 2]
```diff
@@ -31,11 +31,11 @@
 
   private static final Logger LOGGER = LoggerFactory.getLogger(RayServeReplica.class);
 
-  private String backendTag;
+  private String DeploymentTag;
```
[the remark this answers is not in the document]
Comment (jiaodong) [reply java/serve/src/main/java/io/ray/serve/RayServeReplica.java:34]: good call, i noticed a few java changes like this as well and let me ensure they're fixed :)

**Review by edoakes — APPROVED** [review]

**Review by simon-mo — APPROVED** [review]

**Comment by simon-mo** [comment]
@jiaodong before we merge it, can you add back a BackendConfig class so users knows how to migrate? Something like the following (need better messaging though)

```
class BackendConfig:
    def __init__(self, *args, **kwargs):
        raise DeprecationError("This is deprecated, please use DeployementConfig")
```

This should give a better error message, rather than getting a `BackendConfig` not found/not defined and not knowing where to go.

**Comment by jiaodong** [comment]
> @jiaodong before we merge it, can you add back a BackendConfig class so users knows how to migrate? Something like the following (need better messaging though)
> 
> ```
> class BackendConfig:
>     def __init__(self, *args, **kwargs):
>         raise DeprecationError("This is deprecated, please use DeployementConfig")
> ```
> 
> This should give a better error message, rather than getting a `BackendConfig` not found/not defined and not knowing where to go.

good point, i will add it in later commits before all tests are green again.

**Comment by jiaodong** [comment]
the only failed test `test_snapshot.py` is due to timeout, i manually verified they're passing but finished in 119.57s.

**Comment by simon-mo** [comment]
bump the test size?

**Comment by jiaodong** [comment]
`pytest test_cli.py -s` passed for me on local machine but took 146.04s to finish. It seems to be wasting quite a bit of time that struggles to find controller actor that will take a while to show up

```
2021-09-27 09:57:22,840	INFO services.py:1250 -- View the Ray dashboard at http://127.0.0.1:8265
2021-09-27 09:57:33,756	INFO worker.py:826 -- Connecting to existing Ray cluster at address: 192.168.86.154:6379
Traceback (most recent call last):
  File "/Users/jiaodong/Workspace/ray/python/ray/serve/api.py", line 489, in _connect
    controller = ray.get_actor(
  File "/Users/jiaodong/Workspace/ray/python/ray/_private/client_mode_hook.py", line 89, in wrapper
    return func(*args, **kwargs)
  File "/Users/jiaodong/Workspace/ray/python/ray/worker.py", line 1833, in get_actor
    return worker.core_worker.get_named_actor_handle(name, namespace or "")
  File "python/ray/_raylet.pyx", line 1674, in ray._raylet.CoreWorker.get_named_actor_handle
  File "python/ray/_raylet.pyx", line 160, in ray._raylet.check_status
ValueError: Failed to look up actor with name 'SERVE_CONTROLLER_ACTOR'. This could because 1. You are trying to look up a named actor you didn't create. 2. The named actor died. 3. The actor hasn't been created because named actor creation is asynchronous. 4. You did not use a namespace matching the namespace of the actor.
```

File: python/ray/serve/config.py [hunk 3]
```diff
@@ -49,22 +50,29 @@ class AutoscalingConfig(BaseModel):
     # TODO(architkulkarni): Add pydantic validation.  E.g. max_replicas>=min
 
 
-class BackendConfig(BaseModel):
-    """Configuration options for a backend, to be set by the user.
+@Deprecated(error=True)
```
Comment (simon-mo) [inline python/ray/serve/config.py:53]: i think the error will hide the `please use DeployementConfig` migration path.

**Pull request closed without merging**